crates/mun_compiler/src/driver.rs

```diff
@@ -77,15 +86,101 @@ impl Driver {
+        driver.source_root.insert_file(file_id);
+        driver
+            .db
+            .set_source_root(WORKSPACE, Arc::new(driver.source_root.clone()));
 
         Ok((driver, file_id))
     }
+
+    /// Constructs a driver with a package manifest directory
+    pub fn with_package_path<P: AsRef<Path>>(
+        package_path: P,
+        config: Config,
+    ) -> Result<(Package, Driver), anyhow::Error> {
+        // Load the manifest file as a package
+        let package = Package::from_file(package_path)?;
+
+        // Determine output directory
+        let output_dir = ensure_package_output_dir(&package, &config)?;
+
+        // Construct the driver
+        let mut driver = Driver::with_config(config, output_dir)?;
+
+        // Iterate over all files in the source directory of the package and store their information in
+        // the database
+        let source_directory = package
+            .source_directory()
+            .ok_or_else(|| anyhow::anyhow!("the source directory does not exist"))?;
+
+        for source_file_path in iter_source_files(&source_directory) {
+            let relative_path = compute_source_relative_path(&source_directory, &source_file_path)?;
+
+            // Load the contents of the file
+            let file_contents = std::fs::read_to_string(&source_file_path).map_err(|e| {
+                anyhow::anyhow!(
+                    "could not read contents of '{}': {}",
+                    source_file_path.display(),
+                    e
+                )
+            })?;
+
+            let file_id = driver.alloc_file_id(&relative_path)?;
+            driver
+                .db
+                .set_file_relative_path(file_id, relative_path.clone());
+            driver.db.set_file_text(file_id, Arc::new(file_contents));
+            driver.db.set_file_source_root(file_id, WORKSPACE);
+            driver.source_root.insert_file(file_id);
+        }
+
+        // Store the source root in the database
+        driver
+            .db
+            .set_source_root(WORKSPACE, Arc::new(driver.source_root.clone()));
+
+        Ok((package, driver))
+    }
+}
+
+impl Driver {
+    /// Returns a file id for the file with the given `relative_path`. This function reuses FileId's
+    /// for paths to keep the cache as valid as possible.
+    ///
+    /// The allocation of an id might fail if more than file id's exist than can be allocated.
+    pub fn alloc_file_id<P: AsRef<RelativePath>>(
+        &mut self,
+        relative_path: P,
+    ) -> Result<FileId, anyhow::Error> {
+        // Re-use existing id to get better caching performance
+        if let Some(id) = self.path_to_file_id.get(relative_path.as_ref()) {
+            return Ok(*id);
+        }
+
+        // Allocate a new id
+        // TODO: See if we can figure out if the compiler cleared the cache of a certain file, at
+        //  which point we can sort of reset the `next_file_id`
```

Comment:
This is correct because "which" belongs to the previous line (required because otherwise the IDE will not see the entire string as a TODO (only the first part)).